File: web/packages/alloy-audio/src/dsp/effects/phaser.ts
// Multi-stage allpass phaser — per channel, a chain of `stages` first-order
// allpass filters sharing one swept coefficient, plus feedback from the
// chain's last output. The coefficient (tan/pow-heavy) is recomputed only
// once per EFFECT_CONTROL_INTERVAL samples (control rate); the allpass
// chain itself runs every sample (full rate) — same two-rate philosophy as
// the voice's TVF/LFO (see voice.ts CONTROL_INTERVAL). Twin: Phaser.swift.

import { TWO_PI } from '../dsp-types.js';
import { EFFECT_CONTROL_INTERVAL, type EffectUnit, type PhaserParams } from './effect-types.js';

/** Sweep range for the shared allpass coefficient's underlying cutoff. */
export const PHASER_F_MIN = 200;
export const PHASER_F_MAX = 2200;

const OFFSET_L = 0;
const OFFSET_R = 0.25;

export class Phaser implements EffectUnit {
  private readonly zL: Float64Array;
  private readonly zR: Float64Array;
  private lastOutL = 0;
  private lastOutR = 0;
  private phase = 0;
  private sampleCounter = 0;
  private coefL = 0;
  private coefR = 0;

  constructor(
    private readonly params: PhaserParams,
    private readonly sampleRate: number,
  ) {
    this.zL = new Float64Array(params.stages);
    this.zR = new Float64Array(params.stages);
    this.updateCoefficients();
  }

  reset(): void {
    this.zL.fill(0);
    this.zR.fill(0);
    this.lastOutL = 0;
    this.lastOutR = 0;
    this.phase = 0;
    this.sampleCounter = 0;
    this.updateCoefficients();
  }

  process(left: Float32Array, right: Float32Array, frames: number): void {
    const { stages, feedback, mix } = this.params;
    const { zL, zR } = this;

    for (let i = 0; i < frames; i++) {
      if (this.sampleCounter % EFFECT_CONTROL_INTERVAL === 0) {
        this.updateCoefficients();
      }

      const l = left[i];
      const r = right[i];

      let xl = l + this.lastOutL * feedback;
      for (let s = 0; s < stages; s++) {
        const y = -this.coefL * xl + zL[s];
        zL[s] = xl + this.coefL * y;
        xl = y;
      }
      this.lastOutL = xl;

      let xr = r + this.lastOutR * feedback;
      for (let s = 0; s < stages; s++) {
        const y = -this.coefR * xr + zR[s];
        zR[s] = xr + this.coefR * y;
        xr = y;
      }
      this.lastOutR = xr;

      left[i] = l * (1 - mix) + xl * mix;
      right[i] = r * (1 - mix) + xr * mix;

      this.phase += this.params.rateHz / this.sampleRate;
      this.phase -= Math.floor(this.phase);
      this.sampleCounter++;
    }
  }

  private updateCoefficients(): void {
    const { depth } = this.params;

    const sweepL = 0.5 + 0.5 * depth * Math.sin(TWO_PI * (this.phase + OFFSET_L));
    const fL = PHASER_F_MIN * (PHASER_F_MAX / PHASER_F_MIN) ** sweepL;
    const tL = Math.tan((Math.PI * fL) / this.sampleRate);
    this.coefL = (tL - 1) / (tL + 1);

    const sweepR = 0.5 + 0.5 * depth * Math.sin(TWO_PI * (this.phase + OFFSET_R));
    const fR = PHASER_F_MIN * (PHASER_F_MAX / PHASER_F_MIN) ** sweepR;
    const tR = Math.tan((Math.PI * fR) / this.sampleRate);
    this.coefR = (tR - 1) / (tR + 1);
  }
}

File: swift/Sources/AlloyAudio/DSP/Effects/Phaser.swift
import Foundation

/// Multi-stage allpass phaser — per channel, a chain of `stages` first-order
/// allpass filters sharing one swept coefficient, plus feedback from the
/// chain's last output. The coefficient (tan/pow-heavy) is recomputed only
/// once per EffectConstants.controlInterval samples (control rate); the
/// allpass chain itself runs every sample (full rate) — same two-rate
/// philosophy as the voice's TVF/LFO (see Voice.swift CONTROL_INTERVAL).
/// Twin of web src/dsp/effects/phaser.ts (canonical).

/// Sweep range for the shared allpass coefficient's underlying cutoff.
public let PHASER_F_MIN = 200.0
public let PHASER_F_MAX = 2200.0

private let offsetL = 0.0
private let offsetR = 0.25

public final class Phaser: EffectUnit {
    private let params: PhaserParams
    private let sampleRate: Double
    private var zL: [Double]
    private var zR: [Double]
    private var lastOutL = 0.0
    private var lastOutR = 0.0
    private var phase = 0.0
    private var sampleCounter = 0
    private var coefL = 0.0
    private var coefR = 0.0

    public init(params: PhaserParams, sampleRate: Double) {
        self.params = params
        self.sampleRate = sampleRate
        self.zL = [Double](repeating: 0, count: params.stages)
        self.zR = [Double](repeating: 0, count: params.stages)
        updateCoefficients()
    }

    public func reset() {
        for i in 0..<zL.count { zL[i] = 0 }
        for i in 0..<zR.count { zR[i] = 0 }
        lastOutL = 0
        lastOutR = 0
        phase = 0
        sampleCounter = 0
        updateCoefficients()
    }

    public func process(left: inout [Float], right: inout [Float], frames: Int) {
        let stages = params.stages
        let feedback = params.feedback
        let mix = params.mix

        for i in 0..<frames {
            if sampleCounter % EffectConstants.controlInterval == 0 {
                updateCoefficients()
            }

            let l = Double(left[i])
            let r = Double(right[i])

            var xl = l + lastOutL * feedback
            for s in 0..<stages {
                let y = -coefL * xl + zL[s]
                zL[s] = xl + coefL * y
                xl = y
            }
            lastOutL = xl

            var xr = r + lastOutR * feedback
            for s in 0..<stages {
                let y = -coefR * xr + zR[s]
                zR[s] = xr + coefR * y
                xr = y
            }
            lastOutR = xr

            left[i] = Float(l * (1 - mix) + xl * mix)
            right[i] = Float(r * (1 - mix) + xr * mix)

            phase += params.rateHz / sampleRate
            phase -= phase.rounded(.down)
            sampleCounter += 1
        }
    }

    private func updateCoefficients() {
        let depth = params.depth

        let sweepL = 0.5 + 0.5 * depth * sin(DspConstants.twoPi * (phase + offsetL))
        let fL = PHASER_F_MIN * pow(PHASER_F_MAX / PHASER_F_MIN, sweepL)
        let tL = tan(Double.pi * fL / sampleRate)
        coefL = (tL - 1) / (tL + 1)

        let sweepR = 0.5 + 0.5 * depth * sin(DspConstants.twoPi * (phase + offsetR))
        let fR = PHASER_F_MIN * pow(PHASER_F_MAX / PHASER_F_MIN, sweepR)
        let tR = tan(Double.pi * fR / sampleRate)
        coefR = (tR - 1) / (tR + 1)
    }
}
